org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/editor/XbaseEditorInputRedirector.java

```diff
@@ -0,0 +1,162 @@
+
+	public ITypeRoot getTypeRoot(IEditorInput editorInput) {
+		IJavaElement adapter = editorInput.getAdapter(IJavaElement.class);
+		return adapter instanceof ITypeRoot ? (ITypeRoot) adapter : null;
+	}
+
+	/**
+	 * @return the original source for an editor input that points to an Xtext resource copied to the output folder, the given input
+	 *         otherwise
+	 */
+	public IEditorInput findOriginalSourceForOuputFolderCopy(IEditorInput input) {
+		try {
+			IFile resource = ResourceUtil.getFile(input);
+			if (resource != null) {
+				// if the given resource is already pointing to a language file
+				if (fileExtensionProvider.isValid(resource.getFullPath().getFileExtension())) {
+					IJavaProject project = JavaCore.create(resource.getProject());
+					if (project.exists()) {
+						// Check if it's sitting in the output dir, look for original one
+						if (project.getOutputLocation().isPrefixOf(resource.getFullPath())) {
+							IPath relative = resource.getFullPath().removeFirstSegments(project.getOutputLocation().segmentCount());
+
+							List<IPackageFragmentRoot> packageFragmentRoots = new ArrayList<>();
+							for (IPackageFragmentRoot root : project.getPackageFragmentRoots()) {
+								if (root.getKind() == IPackageFragmentRoot.K_SOURCE) {
+									packageFragmentRoots.add(root);
+								}
+							}
+							
+							for (IPackageFragmentRoot source : packageFragmentRoots) {
+								IPath fullPath = source.getCorrespondingResource().getProjectRelativePath().append(relative);
+								IFile newFile = resource.getProject().getFile(fullPath);
+								if (newFile.exists()) {
+									return new FileEditorInput(newFile);
+								}
+							}
+						}
+
+						// Check if it's sitting in one of the output folders set on the source folders
+						List<IClasspathEntry> classpathEntries = new ArrayList<>();
+						for (IClasspathEntry classpathEntry : project.getRawClasspath()) {
+							if (classpathEntry.getEntryKind() == IClasspathEntry.CPE_SOURCE) {
+								classpathEntries.add(classpathEntry);
+							}
+						}
+						
+						for (IClasspathEntry sourceFolder : classpathEntries) {
+							if (sourceFolder.getOutputLocation() != null && sourceFolder.getOutputLocation().isPrefixOf(resource.getFullPath())) {
+								IPath relative = resource.getFullPath().removeFirstSegments(sourceFolder.getOutputLocation().segmentCount());
+								Optional<IPackageFragmentRoot> source = FluentIterable.from(project.findPackageFragmentRoots(sourceFolder))
+										.first();
+								if (source.isPresent()) {
+									IPath fullPath = source.get().getCorrespondingResource().getProjectRelativePath().append(relative);
+									IFile newFile = resource.getProject().getFile(fullPath);
+									if (newFile.exists()) {
+										return new FileEditorInput(newFile);
+									}
+								}
+							}
+						}
+
+					}
+				}
+			}
+		} catch (JavaModelException e) {
+			LOG.error("Error finding the original source for the editor input", e);
```

Comment:
Logging an error message. The chose to do it after looking at other places in the source code where `JavaModelException` is caught. Please suggest if you would like me to replace with a `RuntimeException`.

Reply:
Going to leave this one for @szarnekow

Reply:
Please only log on error level if the exception is not `e.isDoesNotExist()`

Reply:
@szarnekow What should I do if the condition `e.isDoesNotExist()` is not met?